### PR TITLE
Stage 1 Orchestrator RFC

### PR DESCRIPTION
**Pull request by ferozsalam**
Moving the orchestrator RFC from stage 0 into stage 1

[Preview of markdown proposal](https://github.com/ferozsalam/ecs/blob/stage-1-orchestrator-rfc/rfcs/text/0012-orchestrator-field-set.md)

### CODE REVIEW

**Comment by cyrille-leclerc** [comment]
FYI the Metricbeat is working on this topic in parallel, see https://github.com/elastic/beats/issues/17467

File: rfcs/text/0012-orchestrator-field-set.md
Comment (jsoriano) [inline rfcs/text/0012-orchestrator-field-set.md:75]: What would be the use of this field? This may be too dependant on the implementation. Even in the same orchestrator it may be dependant on the kind of resource (e.g. a Deployment might have as subresources the replicaset, the pod, their statuses...).

File: rfcs/text/0012-orchestrator-field-set.md
Comment (jsoriano) [inline rfcs/text/0012-orchestrator-field-set.md:145]: Could you also include some example for other orchestrator?

File: rfcs/text/0012-orchestrator-field-set.md
Comment (jsoriano) [inline rfcs/text/0012-orchestrator-field-set.md:40]: Make this an object in case it includes more data about the cluster in the future? e.g. it could include the name, some other identifier, the endpoint, the version...
See https://github.com/elastic/beats/issues/17467#issuecomment-610381848

File: rfcs/text/0012-orchestrator-field-set.md
Comment (jsoriano) [inline rfcs/text/0012-orchestrator-field-set.md:49]: Cloud Foundry has `organization`s and `space`s.

Wdyt about having an `organization` field for multi-tenant clusters, and `namespace` for logical separations inside the cluster/organization?

File: rfcs/text/0012-orchestrator-field-set.md
Comment (jsoriano) [inline rfcs/text/0012-orchestrator-field-set.md:178]: :+1:

**Review by jsoriano — COMMENTED** [review]
Thanks for moving this! Added some questions and suggestions.

File: rfcs/text/0012-orchestrator-field-set.md
Comment (kaiyan-sheng) [reply rfcs/text/0012-orchestrator-field-set.md:40]: Agreed! Maybe for your `example: production` here, it can be a more specific field under `cluster` object. For example: `orchestrator.cluster.environment` in this case.

File: rfcs/text/0012-orchestrator-field-set.md
Comment (ebeahan) [reply rfcs/text/0012-orchestrator-field-set.md:66]: Maybe? Do you think we'd see a wide need for users to compare resources from different orchestrator types vs. just finding resources of a type in their particular orchestrator?

File: rfcs/text/0012-orchestrator-field-set.md
Comment (ebeahan) [inline rfcs/text/0012-orchestrator-field-set.md:225]: These references are fantastic! 💯

**Review by ebeahan — COMMENTED** [review]
Thanks for continuing to move this proposal forward, @ferozsalam! Excited to see all the discussions already taking place. 😃 

Would you also update the target stage in the proposal: `Stage: **1 (draft)**`?

File: rfcs/text/0012-orchestrator-field-set.md
Comment (jsoriano) [reply rfcs/text/0012-orchestrator-field-set.md:66]: Yeah, most of the users are probably interested on a particular orchestrator, so probably there is no need to have this set of type-agnostic values.

Another reason to have a known set of values is that this could help to have orchestrator-agnostic dashboards. But such thing is probably not so useful because each orchestrator has its own particularities.

In any case it should be possible to combine resources from different orchestrators taking into account the different values when doing the queries.

So I am ok with continuing without having a fixed list of orchestrator-agnostic values.

File: rfcs/text/0012-orchestrator-field-set.md
Comment (ferozsalam) [reply rfcs/text/0012-orchestrator-field-set.md:75]: Kubernetes subresources are a poorly defined concept - I can't find any good documentation to point to, but see [here](https://kubernetes.io/docs/tasks/extend-kubernetes/custom-resources/custom-resource-definitions/#subresources) and [here](https://stackoverflow.com/questions/49396607/where-can-i-get-a-list-of-kubernetes-api-resources-and-subresources) for some examples. Basically, it's perhaps something that can be done to a resource - technically a `pod` is not a subresource of a `deployment`.

For pods, the specific list of subresources is: `attach`, `binding`, `eviction`, `exec`, `log`, `portforward`, `proxy`, and `status` - at least on a vanilla microk8s installation. You can see the full list for your flavour of k8s by following the instructions in the second link above.

There is an alternative representation of subresources: they can be represented alongside the resources themselves, as `pods/attach` for example. So it's perhaps not essentials that they exist as a separate entity in this fieldset.

I am not sure if the resource/subresource distinction translates at all to other orchestrators, so I'm happy to remove this field if it is too specific.

File: rfcs/text/0012-orchestrator-field-set.md
Comment (ferozsalam) [reply rfcs/text/0012-orchestrator-field-set.md:49]: Sounds reasonable to me, I'll add this 👍

File: rfcs/text/0012-orchestrator-field-set.md
Comment (ferozsalam) [reply rfcs/text/0012-orchestrator-field-set.md:40]: Sounds good to me! Will add this shortly.

File: rfcs/text/0012-orchestrator-field-set.md
Comment (ferozsalam) [reply rfcs/text/0012-orchestrator-field-set.md:145]: Yep - I've added an example for Hashicorp Nomad.

File: rfcs/text/0012-orchestrator-field-set.md
Comment (ferozsalam) [reply rfcs/text/0012-orchestrator-field-set.md]: I have made the update, but a question occurred to me as I was doing so - would it make sense to define the common expected fields that would come under the `cluster` object, or is best practice to keep it as open as possible for flexibility? I've kept it open for now.

File: rfcs/text/0012-orchestrator-field-set.md
Comment (ebeahan) [reply rfcs/text/0012-orchestrator-field-set.md:36]: Yeah, the fields being proposed in the `orchestrator.*` fieldset.

Comment (ferozsalam) [reply rfcs/text/0012-orchestrator-field-set.md]: Done 👍

File: rfcs/text/0012-orchestrator-field-set.md
Comment (jsoriano) [reply rfcs/text/0012-orchestrator-field-set.md:81]: > There is an alternative representation of subresources: they can be represented alongside the resources themselves, as `pods/attach` for example. So it's perhaps not essentials that they exist as a separate entity in this fieldset.

Perhaps to identify events related to subresources we should use specific fields, for example for a pod's container metrics, the `container` fields could be used:
```
orchestrator.type: kubernetes
orchestrator.resource: pod
container.name: somecontainer
```
Same thing might happen with other relevant subresources.

So I think we can go on without `orchestrator.subresource` by now and try to use other fields for relevant subresources. In the future we can still add it if we find a better definition.

File: rfcs/text/0012-orchestrator-field-set.md
Comment (jsoriano) [reply rfcs/text/0012-orchestrator-field-set.md:153]: Thanks.

File: rfcs/text/0012-orchestrator-field-set.md
Comment (jsoriano) [reply rfcs/text/0012-orchestrator-field-set.md]: Yes please, add the common expected fields under the `cluster` object. At least `name` and `url` seem to be relevant.

File: rfcs/text/0012-orchestrator-field-set.md
Comment (jsoriano) [inline rfcs/text/0012-orchestrator-field-set.md:143]:
```suggestion
    },
```

File: rfcs/text/0012-orchestrator-field-set.md
Comment (jsoriano) [inline rfcs/text/0012-orchestrator-field-set.md:178]:
```suggestion
      "orchestrator.namespace": "default",
```

Comment (jsoriano) [inline rfcs/text/0012-orchestrator-field-set.md:178]: Why are these `orchestrator` fields under `request`? Shouldn't they be at the root level? I would also expect `orchestrator.type: nomad`.

File: rfcs/text/0012-orchestrator-field-set.md
Comment (jsoriano) [inline rfcs/text/0012-orchestrator-field-set.md:46]: Use lowercased keywords. The example should include a real valid example, if you want to mention more possible cases use the description, something like this:
```suggestion
      example: kubernetes
      description: >
        Orchestrator cluster type (e.g. kubernetes, nomad or cloudfoundry).
```

File: rfcs/text/0012-orchestrator-field-set.md
Comment (jsoriano) [inline rfcs/text/0012-orchestrator-field-set.md:144]:
```suggestion
    "orchestrator.type": "kubernetes",
```

File: rfcs/text/0012-orchestrator-field-set.md
Comment (jsoriano) [inline rfcs/text/0012-orchestrator-field-set.md:146]: These values should be in their singular forms.
```suggestion
    "orchestrator.resource.type": "pod",
```

File: rfcs/text/0012-orchestrator-field-set.md
Comment (kaiyan-sheng) [inline rfcs/text/0012-orchestrator-field-set.md:37]: I like the flexibility we have here for setting `orchestrator.cluster` as an object. Do you think we should also define several common fields under cluster here? For example `orchestrator.cluster.name`? cc: @jsoriano @ChrsMark

Comment (jsoriano) [reply rfcs/text/0012-orchestrator-field-set.md:37]: Yes, I'd say that the most relevant ones should be defined, at least `cluster.name` and `cluster.url`.

Comment (ferozsalam) [reply rfcs/text/0012-orchestrator-field-set.md:37]: 👋 Thanks, I've added the `cluster.name` and `cluster.url` fields. I can't think of anything more that would be essential here at the moment, but happy to add more fields if there are any suggestions.

**Comment by ferozsalam** [comment]
@ebeahan I'm slightly confused by the latest build failures, which I _think_ are caused by something in the test framework that needs updating. The error refers to a Golang file that I haven't touched:

```@@ -19,6 +19,7 @@ go.uber.org/zap v1.9.1/go.mod h1:vwi/ZaCAaUcBkycHslxD9B2zi4UTXhF60s6SWpuDF0Q=
 golang.org/x/sys v0.0.0-20181217223516-dcdaa6325bcb h1:zzdd4xkMwu/GRxhSUJaCPh4/jil9kAbsU7AUmXboO+A=
 golang.org/x/sys v0.0.0-20181217223516-dcdaa6325bcb/go.mod h1:STP8DvDyc/dI5b8T5hshtkjS+E42TnysNCUPdjciGhY=
 gopkg.in/check.v1 v0.0.0-20161208181325-20d25e280405/go.mod h1:Co6ibVJAznAaIkqp8huTwlJQCZ016jof/cbN4VW5Yz0=
+gopkg.in/check.v1 v1.0.0-20180628173108-788fd7840127 h1:qIbj1fsPNlZgppZ+VLlY7N33q108Sa+fhmuc+sWQYwY=
 gopkg.in/check.v1 v1.0.0-20180628173108-788fd7840127/go.mod h1:Co6ibVJAznAaIkqp8huTwlJQCZ016jof/cbN4VW5Yz0=
 gopkg.in/yaml.v2 v2.2.2 h1:ZCJp+EgiOT7lHqUV2J862kp8Qj64Jo6az82+3Td9dZw=
 gopkg.in/yaml.v2 v2.2.2/go.mod h1:hI93XBmqTisBFMUTm0b8Fm+jr3Dg1NNxqwp+5A1VGuI=
git update-index --refresh
scripts/go.sum: needs update
make: *** [Makefile:25: check] Error 1
Error: Process completed with exit code 2.
```

This branch isn't missing any commits that are on master, and I've only touched the RFC Markdown, so I'm stumped. Any ideas?

File: rfcs/text/0012-orchestrator-field-set.md
Comment (ChrsMark) [reply rfcs/text/0012-orchestrator-field-set.md:37]: ++

File: rfcs/text/0012-orchestrator-field-set.md
Comment (ChrsMark) [inline rfcs/text/0012-orchestrator-field-set.md:88]: how about adding `orchestrator.version` too? I see value to know what version of Kubernetes cluster I run on (ie 1.18 vs 1.19 etc)

**Comment by ebeahan** [comment]
> @ebeahan I'm slightly confused by the latest build failures, which I think are caused by something in the test framework that needs updating.

@ferozsalam Thanks for the mention, it's being investigated. The pass/fail isn't applicable to any changes made here, so don't worry about the test failures.

File: rfcs/text/0012-orchestrator-field-set.md
Comment (kaiyan-sheng) [reply rfcs/text/0012-orchestrator-field-set.md:37]: @ferozsalam Thank you!!

File: rfcs/text/0012-orchestrator-field-set.md
Comment (kaiyan-sheng) [reply rfcs/text/0012-orchestrator-field-set.md:88]: Good point @ChrsMark! Or maybe this field can also go under `orchestrator.cluster` to be `orchestrator.cluster.version`?

File: rfcs/text/0012-orchestrator-field-set.md
Comment (jsoriano) [inline rfcs/text/0012-orchestrator-field-set.md:51]: I would suggest to remove this empty `cluster` object. It would be enough like this:
```suggestion
    - name: cluster.name
      level: extended
      type: keyword
      description: >
        Name of the cluster.

    - name: cluster.url
      level: extended
      type: keyword
      description: >
        URL of the cluster.
```

If we want to leave this to document the `cluster` object itself, we could place `name` as `url` as fields of this object.
```suggestion
    - name: cluster
      level: extended
      type: object
      description: >
        Orchestrator cluster details.
      fields:
      - name: name
        level: extended
        type: keyword
        description: >
          Name of the cluster.

      - name: url
        level: extended
        type: keyword
        description: >
          URL of the cluster.
```

File: rfcs/text/0012-orchestrator-field-set.md
Comment (kaiyan-sheng) [reply rfcs/text/0012-orchestrator-field-set.md:88]: Ahh maybe we should keep `orchestrator.version` at the same level as `orchestrator.type`. I don't have a strong opinion here :)

Comment (ferozsalam) [reply rfcs/text/0012-orchestrator-field-set.md:88]: The `version` field is a good shout - I'll add it under the `cluster` object, as if we're defining that to contain all the cluster-wide fields, `version` definitely fits in that category.

**Review by kaiyan-sheng — DISMISSED** [review]
This looks good to me, thank you!!

**Review by ChrsMark — DISMISSED** [review]

**Review by jsoriano — APPROVED** [review]
Thanks! LGTM

**Review by ebeahan — DISMISSED** [review]
The collaboration here has been fantastic! Thanks to @ferozsalam for all the iterations and to @jsoriano @kaiyan-sheng @ChrsMark for all the feedback.

I noticed one possible minor edit, but otherwise, I think the proposal looks great.

Next steps:

* It's been discussed out-of-band that @exekias will be acting as the sponsor here. I'll add @exekias to the `People` section when I update the advancement date. If this is incorrect, let me know.
* The ECS team has found it useful to create a directory using the RFC number to capture examples and the proposed schema changes as standalone YML files (for use with `--include`). Just informational - no action required; I'll handle copy/pasting the YAML into that new file when I merge the proposal. 
* By advancing to stage 1, the `orchestrator.*` fields will be added to the experimental ECS schema. I'll also handle opening a separate PR for that implementation.
* Once we merge this PR, I suggest creating the stage 2 PR shortly after even if the only change is updating the `stage` at the top of the document. It's great to already have the PR opened when thoughts or feedback come up.

**Pull request merged**
